uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm

```diff
@@ -715,67 +715,219 @@ RETURNS:
+        $scan_type, 'warning', \@warning_headers, $file
+    );
+
+    if (%validWarningFields) {
+        $this->insert_into_mri_violations_log(
+            \%validWarningFields, 'warning', $pname, $candID, $visit_label, $file
+        );
+        return ('warn');
+    }
+
+    ## Step 4 - if we end up here, then the file passes the extra validation
+    # checks and return 'pass'
+    return ('pass');
+}
+
+
+=pod
+
+=head3 loop_through_protocol_violations_checks($scan_type, $severity, $headers, $file)
+
+Loops through all protocol violations checks for a given severity and creates
+a hash with all the checks that need to be applied on that specific scan type
+and severity.
+
+INPUTS:
+  - $scan_type: scan type of the file
+  - $severity : severity of the checks we want to loop through (exclude or warning)
+  - $headers  : list of different headers found in the C<mri_protocol_checks>
+                table for a given scan type
+  - $file     : file information hash ref
+
+RETURNS: a hash with all information about the checks for a given scan type
+and severity
+
+=cut
+
+sub loop_through_protocol_violations_checks {
+    my ($this, $scan_type, $severity, $headers, $file) = @_;
+
+    my %valid_fields; # will store all information about what fails
+
+    # query to fetch list of valid protocols in mri_protocol_checks table
+    my $query = "SELECT * FROM mri_protocol_checks "
+                . "WHERE Scan_type=? AND Severity=? AND Header=?";
+    my $sth   = ${$this->{'dbhr'}}->prepare($query);
+
+    # loop through all severity headers for the scan type and check if in the
+    # value of the header in the file fits one of the valid range present in
+    # mri_protocol_checks
+    foreach my $header (@$headers) {
+        # get the value from the file
+        my $value = $file->getParameter($header);
+
+        # execute query for $scan_type, $severity, $header
+        $sth->execute($scan_type, $severity, $header);
+
+        # grep all valid ranges and regex to compare with value in the file
+        my (@valid_ranges, @valid_regexs);
+        while (my $check = $sth->fetchrow_hashref()) {
+            push(@valid_ranges, $check->{'ValidRange'}) if $check->{'ValidRange'};
+            push(@valid_regexs, $check->{'ValidRegex'}) if $check->{'ValidRegex'};
+        }
+
+        # go to the next header if did not find any checks for that scan
+        # type, severity and header
+        next if (!@valid_ranges && !@valid_regexs);
+
+        # loop through all checks
+        my $is_valid;
+        my (@failed_valid_ranges, @failed_valid_regexs);
+        foreach my $valid_range (@valid_ranges) {
+            if ($valid_range && (NeuroDB::MRI::in_range($value,
+                $valid_range))) {
+                $is_valid = 1;
+            } else {
+                push(@failed_valid_ranges, $valid_range);
+            }
+        }
+        foreach my $valid_regex (@valid_regexs) {
```

Comment:
Same logic here:

```perl
next if grep($value =~ /$_->{'ValidRegex'}/, @valid_regexs);
```

and if we get past this point, we know that @valid_regexs will only contain references to rows for which `$value` does __not__ match the regex for that row.